src/common/providers/database.ts

```diff
@@ -0,0 +1,342 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2022 Firebase
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import * as firebase from 'firebase-admin';
+import { joinPath, /*normalizePath,*/ pathParts } from '../../utilities/path';
+
+/**
+ * Interface representing a Firebase Realtime Database data snapshot.
+ */
+export class DataSnapshot {
+  public instance: string;
+
+  /** @hidden */
+  private _ref: firebase.database.Reference;
+
+  /** @hidden */
+  private _path: string;
+
+  /** @hidden */
+  private _data: any;
+
+  /** @hidden */
+  private _childPath: string;
+
+  constructor(
+    data: any,
+    path?: string, // path will be undefined for the database root
+    private app?: firebase.app.App,
+    instance?: string
+  ) {
+    if (app?.options?.databaseURL?.startsWith('http:')) {
+      // In this case we're dealing with an emulator
+      this.instance = app.options.databaseURL;
+    } else if (instance) {
+      // SDK always supplies instance, but user's unit tests may not
+      this.instance = instance;
+    } else if (app) {
+      this.instance = app.options.databaseURL;
+    } else if (process.env.GCLOUD_PROJECT) {
+      this.instance =
+        'https://' + process.env.GCLOUD_PROJECT + '.firebaseio.com';
```

Comment:
When would we ever reach this? In tests only?

The default RTDB database should be `'https://' + process.env.GCLOUD_PROJECT + '-default-rtdb.firebaseio.com'` if not specified in firebase config.

Reply:
This is left over from v1 and we'd always supply the instance in the SDK. I'll change it to the correct default RTDB url.

Reply:
Thanks!